### ISSUE TITLE
cross-referencing IDREFs on aria-labelledby and aria-controls + button with both aria-haspopup and aria-expanded attributes

### ISSUE THREAD

**Issue by devarshipant** [w3c/aria-practices#476]
Refer to the example - http://w3c.github.io/aria-practices/examples/menu-button/menu-button-actions-active-descendant.html

The button has both aria-haspopup and aria-expanded plus references the role= menu. Isn't aria-haspopup redundant when used with role menu and aria-expanded true? Hearing 'menu' + 'press spacebar and arrow keys to navigate' may be a sufficient for users to know the state of the control. 

Note -- Both NVDA 2017 and JAWS 18 ignore aria-haspopup and aria-expanded attributes.

On the button element, most likely due to the use of aria-controls attribute, JAWS speaks 'Use JAWS key + Alt + M to move to controlled element,' a helpful feature in JAWS (JAWS 18 / IE 11). But that feature does not work. 

<img width="612" alt="howmanyrelattr" src="https://user-images.githubusercontent.com/7377040/31452810-105f6bdc-ae7e-11e7-85db-3a946d2022d7.png">

**Comment by mcking65** [comment]
@devarshipant commented:

> The button has both aria-haspopup and aria-expanded plus references the role= menu. Isn't aria-haspopup redundant when used with role menu and aria-expanded true? 

No; aria-haspopup and aria-expanded serve two different purposes.

> Note -- Both NVDA 2017 and JAWS 18 ignore aria-haspopup and aria-expanded attributes.

Actually they do not, but I can understand why you might think that.

Note that both JAWS and NVDA both call the button a menu button instead of a button. That is because the browser mapping in the accessibility tree is effected by the presence of aria-haspopup. Remove aria-haspopup, and the screen reader sees a button. As explained in the design pattern, aria-pressed has a similar effect.

This situation where properties effect role mappings, in my opinion, is an unfortunate decision that was made way back in the early days of ARIA -- before I joined the working group. When I started paying attention in the 2009 time frame, I certainly didn't forsee the impact it would have. As I recall, the rationale was to avoid role explosion. Instead we have role confusion. Personally, I wish we had a menubutton role and a togglebutton role. We now have a switch role, which is very similar to a togglebutton so there is less reason to use button with aria-pressed in some scenarios. But, if you want an element that is a button that opens a menu, you need to use role button with either aria-haspopup="true" or aria-haspopup="menu".

We only apply aria-expanded in this pattern when the menu is visible. The primary benefit there is for the reading mode and for touch screen readers. If you use this on mobile, and touch the button, you can hear that the menu button is expanded and activating it will collapse the menu.

> On the button element, most likely due to the use of aria-controls attribute, JAWS speaks 'Use JAWS key + Alt + M to move to controlled element,' a helpful feature in JAWS (JAWS 18 / IE 11). But that feature does not work.

The value of aria-controls is admitedly limited. That is mostly due to lack of assistive technology support. And, I am not at all fond of the JAWS support of aria-controls; I turn it off.JAWS and NVDA do not ignore aria-haspopup. They both call the button a menubutton because of the presence of aria-

**Comment by devarshipant** [comment]
> No; aria-haspopup and aria-expanded serve two different purposes.

Of course they serve different purposes. I am skeptical of the need to use aria-haspopup, aria-expanded, and aria-controls together on the button element. In my testing, voiceover and talkback, besides NVDA and JAWS, do not convey aria-expanded (state), only aria-haspopup (vo says, 'shows popup'). Perhaps, and I could be wrong -- there is too much aria traffic in this example.

> Note that both JAWS and NVDA both call the button a menu button instead of a button. That is because the browser mapping in the accessibility tree is effected by the presence of aria-haspopup. Remove aria-haspopup, and the screen reader sees a button. As explained in the design pattern, aria-pressed has a similar effect.

Correct.

> The value of aria-controls is admitedly limited. That is mostly due to lack of assistive technology support. And, I am not at all fond of the JAWS support of aria-controls; I turn it off.

Maybe aria-controls does not lend itself well in this example and is therefore not required.

**Comment by mcking65** [comment]
@devarshipant commented:

> In my testing, voiceover and talkback, besides NVDA and JAWS, do not convey aria-expanded (state), only aria-haspopup (vo says, 'shows popup').

We have aria-expanded on the button only when the menu is displayed.
In MacOS and Win, the focus is inside  the menu when the menu is displayed so those screen readers will not typically see the button in that state unless the user specifically goes looking for it with the reading cursor.
If you do go looking for the button when the menu is displayed, you will hear that expanded is rendered.

On mobile, double tapping the button leaves the reading cursor on the button and double tapping the button is the way to close the menyu if you don't want to choose a menyu item.
So, on mobile it is important to be able to hear when the menu is displayed by reading the button.

> Maybe aria-controls does not lend itself well in this example and is therefore not required.

I agree that aria-controls is not required but do believe that it is appropriate to recommend using it.

One of the reasons for these examples is to give assistive technology developers cannonical examples of appropriate use of ARIA.
The historical lack of such examples is one reason why support for properties like aria-controls remains under utilized by assistive technologies.

With this example, and others where aria-controls is recommended, we can help the AT devs design experiences that are more robust and helpful to the user by exploiting the intent of this attribute.
At this time, what JAWS does with aria-controls is not very helpful; it actually makes the experience less enjoyable and does not provide functionality that is particularly useful.
To be fair, I think the JAWS team did what they did because they were under pressure from sources other than their user base to do something that demonstrated support for aria-controls.
I am confident it will be changed at some point.

I hope this answers your questions.

**Comment by devarshipant** [comment]
> I hope this answers your questions.

It does, @mcking65. thanks!
